File: algoritmos-y-estructuras-de-datos-2/algorithms/sorting/mergeSort.hs
main :: IO ()
main = do
  let res = mergeSort example
  print res
  print ("Is sorted? " ++ show (isSorted res))
    

split :: [a] -> ([a], [a])
split xs = (take mid xs, drop mid xs)
    where mid = div (length xs) 2

merge :: Ord a => [a] -> [a] -> [a]
merge xs [] = xs
merge [] ys = ys
merge (x:xs) (y:ys)
  | x <= y    = x : merge xs (y:ys)
  | otherwise = y : merge (x:xs) ys

mergeSort :: Ord a => [a] -> [a]
mergeSort []  = []
mergeSort [x] = [x]
mergeSort xs  = merge (mergeSort left) (mergeSort right)
  where (left, right) = split xs
  
isSorted :: Ord a => [a] -> Bool
isSorted []  = True
isSorted [x] = True
isSorted (x:y:ys)
  | x <= y    = isSorted (y:ys)
  | otherwise = False
  
example = [1, 0 , 3, 4, 5, -3, 20, 45, 2]

File: algoritmos-y-estructuras-de-datos-2/algorithms/sorting/mergeSort.js
function mergeSort(A) {
    // Base case
    if (A.length < 2) {
        return A;
    }
    
    // Split array in half
    let mid = Math.floor(A.length / 2);
    let left = A.slice(0, mid);
    let right = A.slice(mid);
    
    // Sort both of the sides
    left = mergeSort(left);
    right = mergeSort(right);
    
    // Merge
    let sorted = [];
    while (left.length && right.length) {
        if (left[0] < right[0]) {
            sorted.push(left.shift());
        } else {
            sorted.push(right.shift());
        }
    }
    
    // If one of the sides empties, concat to the end of sorted
    if (left.length) { sorted = [...sorted, ...left] }
    if (right.length) { sorted = [...sorted, ...right] }
    
    return sorted;
}

let A = [1, 0, -4, 25, 32, 7, -2, 1];

console.log(mergeSort(A));
